### PR TITLE
Allow viewing finalized forms in "Ready to send"

### LINKED ISSUE #5612: No way to view a finalized form before sending
#### ODK Collect version

v2023.2 Beta 1

#### Problem description

There is no longer a way to view a finalized form before it is sent. In previous versions, finalized forms would appear in "Edit Saved Form" where they could be edited to view the current data in them. The corresponding new "Drafts" page only shows non-finalized forms (as finalized forms can no longer be edited) and finalized forms only appear in "Ready to send" where they can be sent but not viewed. It feels likely that some workflows might require the ability to review a form before sending.

#### Steps to reproduce the problem

1. Disable auto-send (or disconnect from all networks)
2. Fill a form and finalize

There's now no way to view the finalized form.

#### Expected behavior

There should be a way to view the form before sending.

### PR DESCRIPTION
**Pull request by seadowg**
Closes #5612 

As discussed in the issue, this makes it so clicking on the form in the list opens it for viewing, but still maintains the ability to select it for sending by clicking on the checkbox directly.

I had to rewrite `InstancesUploaderListActivity` somewhat to get this working. The big change was detaching it from the `AppListActivity` hierarchy and then reimplementing selection using the new `MultiSelectViewModel`. I also had to duplicate a bunch of code (from `AppListActivity`) to maintain features like search and sort.

I don't think the code I've ended up with is ideal - I'd liked to have built a new implementation of "Ready to send" from scratch with a goal of ending up with something similar to (and sharing components with) `DeleteBlankFormFragment`, but that would be very involved, and I'd want to take it slowly which doesn't make sense at this point in time. It's very likely we'll do most of that work as part of #5343 as well as I imagine "Ready to send" will end up sharing most of it's code with "Delete saved form".

### What has been done to verify that this works as intended?

New tests, but mostly verified manually. 

#### Why is this the best possible solution? Were any other approaches considered?

Discussed above.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

There's a lack of test coverage for the selection, sort and search parts of these screens in general, and I haven't done much to improve that here so focusing on that for "Ready to send" is important here. I would almost treat this as if it was a new screen. I'm going to mark this as "needs testing" before review as I think we have a high risk of regressions here, and I'd like to deal with them as quickly as possible.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/androidTest/java/org/odk/collect/android/support/pages/ViewFormPage.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package org.odk.collect.android.support.pages;
+
+import androidx.annotation.NonNull;
+
+import org.odk.collect.android.R;
+
+public class ViewFormPage extends Page<ViewFormPage> {
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/pages/ViewFormPage.java]: New class and it's java? Kotlin would be better here.

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormFinalizingTest.kt [hunk 2]
```diff
@@ -21,27 +22,28 @@ class FormFinalizingTest {
...
             .startBlankForm("One Question")
-            .swipeToEndScreen()
-            .clickSaveAsDraft()
-            .assertNumberOfEditableForms(1)
-            .assertNumberOfFinalizedForms(0)
+            .fillOutAndFinalize(FormEntryPage.QuestionAndAnswer("what is your age", "52"))
+            .assertNumberOfEditableForms(0)
+            .clickSendFinalizedForm(1)
+            .clickOnForm("One Question")
+            .assertText("52")
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormFinalizingTest.kt]: I don't think those three last steps are needed here. This test is for checking if the form is finalized and asserting it ended up in the list of finalized forms is enough. Instead, the last step should be: `assertNumberOfFinalizedForms(1)`

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java [hunk 3]
```diff
@@ -242,4 +249,29 @@ public void onClick(DialogInterface dialog, int i) {
...
+            case BY_NAME_DESC:
+                sortingOrder = DatabaseInstanceColumns.DISPLAY_NAME + " COLLATE NOCASE DESC, " + DatabaseInstanceColumns.STATUS + " DESC";
+                break;
+            case BY_DATE_ASC:
+                sortingOrder = DatabaseInstanceColumns.LAST_STATUS_CHANGE_DATE + " ASC";
+                break;
+            case BY_DATE_DESC:
+                sortingOrder = DatabaseInstanceColumns.LAST_STATUS_CHANGE_DATE + " DESC";
+                break;
+            case BY_STATUS_ASC:
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java]: We do not support sorting by status. It was possible in the past when for example we had incomplete and finalized forms on the same list. So it looks like we can get rid of it.

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderListActivity.java [hunk 4]
```diff
@@ -388,4 +506,61 @@ private boolean showSentAndUnsentChoices() {
...
+            case BY_NAME_DESC:
+                sortingOrder = DatabaseInstanceColumns.DISPLAY_NAME + " COLLATE NOCASE DESC, " + DatabaseInstanceColumns.STATUS + " DESC";
+                break;
+            case BY_DATE_ASC:
+                sortingOrder = DatabaseInstanceColumns.LAST_STATUS_CHANGE_DATE + " ASC";
+                break;
+            case BY_DATE_DESC:
+                sortingOrder = DatabaseInstanceColumns.LAST_STATUS_CHANGE_DATE + " DESC";
+                break;
+            case BY_STATUS_ASC:
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderListActivity.java]: Again sorting by status seems to be redundant.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/androidTest/java/org/odk/collect/android/support/pages/ViewFormPage.java (hunk 1 above)
Comment (seadowg) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/pages/ViewFormPage.java]: Great catch! Not sure what I was thinking 😆.

**Comment by dbemke** [comment]
Forms get deselected in "Ready to send" after minimizing the app when "don't keep activities" is enabled. The issue isn't present on the master version

**Steps to reproduce:**
1. In device settings enabled "don't keep activities".
2. In collect settings turn off the auto send option.
3. Send a submission.
4. Go to "Ready to send" and select the form.
5. Minimize the app.
6. Go back to the app.

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormFinalizingTest.kt (hunk 2 above)
Comment (seadowg) [reply collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormFinalizingTest.kt]: The `clickOnForm` and `assertText` at the end there are what drives out being able to click on forms in this view. Would you prefer that I split that out to another test like `canViewFinalizedForms()`?

Comment (grzesiek2010) [reply collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormFinalizingTest.kt]: Yeah, definitely there should be a separate test for checking if opening finalized forms for review is possible. This test is about saving only.

**Comment by seadowg** [comment]
@dbemke fixed!

**Comment by seadowg** [comment]
@dbemke looks like my solution for saving state when "Don't kill Activities" is on is breaking "Delete blank form". I think there's some deeper work to do there that's not worth it right now. I'm going to just revert that change, and then you can file an issue after we merge this that "Ready to send" and "Delete blank form" do not keep selections when multitasking with "Don't keep activities".

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormFinalizingTest.kt [hunk 5]
```diff
@@ -21,27 +22,26 @@ class FormFinalizingTest {
...
             .copyForm(FORM)
             .assertNumberOfFinalizedForms(0)
             .startBlankForm("One Question")
-            .swipeToEndScreen()
-            .clickSaveAsDraft()
-            .assertNumberOfEditableForms(1)
-            .assertNumberOfFinalizedForms(0)
+            .fillOutAndFinalize(FormEntryPage.QuestionAndAnswer("what is your age", "52"))
+            .assertNumberOfEditableForms(0)
+            .clickSendFinalizedForm(1)
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormFinalizingTest.kt]: There is no need to click on the button just use `assertNumberOfFinalizedForms` like in the test below.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

**Comment by srujner** [comment]
**Tested with Success!**

Verified on device with Android 13

Verified cases:
- Deleting a form from file directory while the form is opened in Ready to send
- Trying to open a deleted form
- Adding instances to Ready to send manually in file directory
- Deleting forms with auto send off and failed submissions in Delete form
- Don't keep activities enabled/disabled
- Changing view to Show Sent/Unsent Forms
- Viewing Finalized Forms
- Minimize the app and rotate the screen
- No background processing option turned off/on

**Comment by dbemke** [comment]
**Tested with Success!**

Verified on device with Android 10

**Review by grzesiek2010 — APPROVED** [review]

**Pull request merged**